lambdaclass/mirra_backend
# Add level progression

Allows tracking user progress in campaigns. Users progress through campaigns when they beat levels.

Campaigns have been wrapped in enclosing containers we called "SuperCampaigns". Reason is we will want to have many modes (think, one story where you are the good guys and one where you are the bad guys), and each one of them will have its **sequential** campaigns with their levels. Each of those stories would be a SuperCampaign. This allows for cross-campaign progression `(1-19 -> 1-20 -> 2-1)`.

# Known issues
If we send a call to play the first level of the campaign while we are at a further level, and beat it, we'll advance the level by beating an easy opponent. [Issue has been created.](https://github.com/lambdaclass/champions_of_mirra/issues/214)

CampaignProgress should really be called SuperCampaignProgresss since we have 1 entry for each. [Issue has been created.](https://github.com/lambdaclass/champions_of_mirra/issues/193)

# To test

1. Create User
`{:ok, user} = Champions.Users.register("User")`
2. Add units
`team1 = Enum.map(1..6, fn slot -> GameBackend.Units.insert_unit(%{character_id: GameBackend.Units.Characters.get_character_by_name("Muflus").id, level: 99999, tier: 1, rank: 1, selected: true, user_id: user.id, slot: slot}) |> elem(1) |> GameBackend.Repo.preload(character: [:basic_skill, :ultimate_skill]) end)`
3. Get campaigns
`[campaign_1, campaign_2 | _] = Champions.Campaigns.get_campaigns`
4. Get levels
`[level1, level2, level3, level4, level5, level6, level7, level8, level9, level10] = campaign_1.levels`
5. For each level, fight it, win it and check that the CampaignProgress for user has changed to the next level.
`Champions.Battle.fight_level(user.id, level1.id)`
`GameBackend.Campaigns.get_campaign_progress user_1.id, campaign.id`
When you beat level 10, the last of the campaign, CampaignProgress should change level _and_ campaign
